cpp/src/neighbors/detail/quantization.cuh

```diff
@@ -0,0 +1,164 @@
+    raft::resource::sync_stream(res);
+
+    // persist settings in params
+    params.min         = double(minmax_h[0]);
+    params.max         = double(minmax_h[1]);
+    params.scalar      = double(range_q_type) / (params.max - params.min + 1.0);
+    params.is_computed = true;
+  }
+
+  // allocate target
+  auto out = raft::make_device_matrix<QuantI, int64_t>(res, dataset.extent(0), dataset.extent(1));
+
+  // raft unary op or raft::linalg::map?
+  // TempT / TempI as intermediate types
+  raft::linalg::unaryOp(out.data_handle(),
+                        dataset.data_handle(),
+                        n_elements,
+                        raft::compose_op(
+                          raft::cast_op<QuantI>{},
+                          raft::add_const_op<int>(q_type_min),
+                          [] __device__(TempI a) {
+                            return raft::max<TempI>(raft::min<TempI>(a, q_type_max - q_type_min),
+                                                    TempI(0));
+                          },
+                          raft::cast_op<TempI>{},
+                          raft::add_const_op<TempT>(0.5),  // for rounding
+                          raft::mul_const_op<TempT>(params.scalar),
+                          raft::sub_const_op<TempT>(params.min),
+                          raft::cast_op<TempT>{}),
+                        stream);
+
+  return out;
+}
+
+template <typename T, typename QuantI = int8_t, typename TempT = double, typename TempI = int64_t>
+raft::host_matrix<QuantI, int64_t> scalar_quantize(raft::resources const& res,
+                                                   cuvs::neighbors::quantization::params& params,
+                                                   raft::host_matrix_view<const T, int64_t> dataset)
+{
+  constexpr TempI q_type_min   = static_cast<TempI>(std::numeric_limits<QuantI>::min());
+  constexpr TempI q_type_max   = static_cast<TempI>(std::numeric_limits<QuantI>::max());
+  constexpr TempI range_q_type = q_type_max - q_type_min + TempI(1);
+
+  size_t n_elements = dataset.extent(0) * dataset.extent(1);
+
+  // conditional: search for quantiles / min / max
+  if (!params.is_computed) {
+    ASSERT(params.quantile > 0.5 && params.quantile <= 1.0,
+           "quantile for scalar quantization needs to be within (.5, 1]");
+
+    double quantile_inv = 1.0 / params.quantile;
+
+    // select subsample
+    int seed = 137;
+    std::mt19937 rng(seed);
+    constexpr size_t num_samples = 10000;
+    size_t subset_size           = std::min(num_samples, n_elements);
+    std::vector<T> subset(subset_size);
+    std::sample(dataset.data_handle(),
+                dataset.data_handle() + n_elements,
+                std::back_inserter(subset),
+                subset_size,
+                rng);
+
+    // quantile / sort and pick for now
+    thrust::sort(thrust::omp::par, subset.data(), subset.data() + subset_size);
```

Comment:
Need to confirm whether sorting on CPU side would be still preferred or not when we increase num_samples.

Reply:
Alternative would be to gather the points on the CPU side, and then call the GPU to do the quantization. Assuming we have a streaming quantile estimation, that would not be faster to copy data to the device. But with the current sorting approach, it is not clear to me what is the better solution.